### PR TITLE
Use discovery heuristics in ChooseHostInterface method to fetch PublicAddress

### PR DESCRIPTION
**Pull request by sub-mod**
Use a better logic to fetch interface for PublicAddress.
ChooseHostInterface will check if the nic is up and will ignore docker0.

### CODE REVIEW

**Comment by sub-mod** [comment]
@thockin  part of https://github.com/GoogleCloudPlatform/kubernetes/issues/3681

**Comment by sub-mod** [comment]
I found out why the integration test is failing after i figured out the 2 nics for the travis-ci setup .
1)interfaces = lo 
IP =127.0.0.1/8  UP=true ,  !loopback=false
2)interfaces = venet0 
IP =127.0.0.2/32 UP=true  !loopback=false
both are loopback for more info look here http://tools.ietf.org/html/rfc3330#127.0.0.1/32

The logic I have is ` if (nic.IsUP && !nic.Loopback) select nic;`
Since both are loopback it fails.

**Comment by sub-mod** [comment]
@thockin apologies for all the commits ,But Had to figure out what was the network config for the test which failed.

File: pkg/master/master.go [hunk 1]
```diff
@@ -197,29 +197,16 @@ func setDefaults(c *Config) {
 		// Find and use the first non-loopback address.
 		// TODO: potentially it'd be useful to skip the docker interface if it
 		// somehow is first in the list.
-		addrs, err := net.InterfaceAddrs()
+		hostIP, found, err := util.ChooseHostInterface([]string{"docker0"})
```
Comment (brendandburns) [inline pkg/master/master.go]: using util.StringSet instead of a list will clean up some of the checking code (eliminate interfaceSkip, for example)

**Comment by thockin** [comment]
Can you take a look at #4218 - they are the same issue on different components - both trying to choose the best public IP.

I think we have a better answer than this - I pasted a play link in #4218 that looks at the route tables to figure out which NIC hosts the default gateway.  PTAL?  If we think that is good enough to try, I or you can patch that in and see what Travis has to say about it...

**Comment by sub-mod** [comment]
@thockin thanks
I made few modifications , Can you please check this. http://play.golang.org/p/X4HZGnwd_M
Are you using "route -n" command ? I am not able to see the same routing fields which you see in your code.Which OS are you using ?
"route -n" returns same output on F21/Ubuntu14/Centos7/CoreOS/RHEL Atomic

#4218 talks only proxier.go code this problem exists in apiserver too.
I think that it is better if master.go and proxier.go uses ChooseHostInterface method to pick a NIC.
Although we have removed the randomness by choosing the nic with gateway everytime ,we still have to give override feature so that traffic can be just confined to a single subnet.

I will test locally and in travis to see if it breaks anything.
btw mostly the nic with gateway is the public IP of the system.Do we really have to pick the public IP nic ?
The automatic discovery seems kinda difficult.For example if there exists a alias IP on eth0 , the code picked aliasIP as the publicAddress but later the alias IP can be easily removed with system restart.And then a new IP would be picked by the daemon.So I feel having --bind-to as an interface name helps in the clarity.

**Comment by thockin** [comment]
The file input was /proc/net/route - should be stable across kernel
versions and distros.

I agree that we should use the same heuristic to choose the default public
interface in both master and proxy.  Before we invest too heavily in coding
this up I wanted to get a first-pass sanity check from people.  I feel like
this will be more stable overall, but I am a bit worried about it is true
multi-NIC environments (e.g. one internal NIC and one public NIC - will we
end up choosing the public?)  I guess that is what the override is for.

**Comment by sub-mod** [comment]
ok i need to check if /proc/net/route  is there on rh-atomic host

On my k8 cluster in the cloud, right now the code picks the public NIC.Because nic0 is used for public acess and other nics are used as management subnet, hadoop cluster subnet etc.I think this should not happen.
So What i need now is override feature --bind-to to select the k8-subnet of my choice, and the public Address should belong to this nic belonging to k8-subnet.(which depending on situations could be anything)

**Comment by thockin** [comment]
What cloud?

**Comment by sub-mod** [comment]
So we have 2 issues here
1)" common way for all kube daemons to be configured , for ex using --bind-to "
https://github.com/GoogleCloudPlatform/kubernetes/issues/3861
which could be a NIC name/IP address/Subnet
it has an implication that if the interface changes IP the daemon will adapt, whcih maynot be true.
for example as mentioned  by @thockin if a host uses DHCP  and you --bind-to-interface=eth0, and then eth0 gets a new IP. 
But i assume that if interface changes IP, we need to do a restart host and then the new change are reflected across all nodes

2)discovery heuristics

Figure out which IP to use incase of multi nic cluster.
option1 : choose the nic with gateway
In worst case : take all the available nics and start removing from them bridge nics , p2p nics , alias nic ,docker nic , virb0 nic...etc. then one among the rest of the candidate nic can be used

**Comment by sub-mod** [comment]
@thockin can you please take a look

**Comment by sub-mod** [comment]
@thockin @brendanburns can you please take a look

File: pkg/util/util.go [hunk 2]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+
+func getRoutes(file io.Reader) ([]Route, error) {
+	routes := []Route{}
+	scanner := bufio.NewReader(file)
+	for {
+		line, err := scanner.ReadString('\n')
+		if strings.HasPrefix(line, "Destination") || strings.HasPrefix(line, "Kernel") {
+			continue
+		}
+		if err == io.EOF {
```
Comment (thockin) [inline pkg/util/util.go]: Prefer to check the error before looking at the result.  I.e. flip these 2 blocks.

File: pkg/util/util.go [hunk 3]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+func getRouteTables() (string, error) {
+	args := []string{"-n"}
+	out, err := exec.Command("route", args...).Output()
+	if err != nil {
+		return "", err
+	}
+	return string(out), err
+}
+
+func ChooseHostInterface() (net.IP, bool, error) {
```
Comment (thockin) [inline pkg/util/util.go]: This needs a nice detailed comment.

File: pkg/util/util.go [hunk 4]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+			return nil, fmt.Errorf("Unable to Parse IP: %s", fields[1])
+		}
+		route.Gateway = ip
+	}
+	return routes, nil
+}
+
+func getRouteTables() (string, error) {
+	args := []string{"-n"}
+	out, err := exec.Command("route", args...).Output()
```
Comment (thockin) [inline pkg/util/util.go]: I'm a little afraid of the potential size of this, which is why I was using a proc file.  Am I just being paranoid?  I'm willing to be convinced...

Comment (thockin) [inline pkg/util/util.go]: yeah, I'm just being dumb.

File: pkg/util/util.go [hunk 5]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+}
+
+func ChooseHostInterface() (net.IP, bool, error) {
+	found := false
+	routedata, err := getRouteTables()
+	if err != nil {
+		fmt.Printf("ERROR:getRouteTables,  %v", err)
+		return nil, false, err
+	}
+	r := strings.NewReader(routedata)
```
Comment (thockin) [inline pkg/util/util.go]: nit: move inline with call to getRoutes() below?

File: pkg/util/util.go [hunk 6]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+	finalIP := net.ParseIP("127.0.0.1")
+	if finalIP == nil {
+		return nil, false, nil
+	}
+	routes, err := getRoutes(r)
+	if err != nil {
+		glog.Infof("ERROR:getRoutes, %v", err)
```
Comment (thockin) [inline pkg/util/util.go]: no need to glog here, just return

File: pkg/util/util.go [hunk 7]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+
+func ChooseHostInterface() (net.IP, bool, error) {
+	found := false
+	routedata, err := getRouteTables()
+	if err != nil {
+		fmt.Printf("ERROR:getRouteTables,  %v", err)
+		return nil, false, err
+	}
+	r := strings.NewReader(routedata)
+	finalIP := net.ParseIP("127.0.0.1")
```
Comment (thockin) [inline pkg/util/util.go]: Why are you parsing an IP only to throw it away later?

File: pkg/util/util.go [hunk 8]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+	}
+	routes, err := getRoutes(r)
+	if err != nil {
+		glog.Infof("ERROR:getRoutes, %v", err)
+		return nil, false, err
+	}
+	for i := range routes {
+		zero := net.IP{0, 0, 0, 0}
+		if routes[i].Destination.Equal(zero) {
+			glog.Infof("Default route transits interface %q\n", routes[i].Interface)
```
Comment (thockin) [inline pkg/util/util.go]: make this glog.V(3).Infof() - I don't think it is generally useful

File: pkg/util/util.go [hunk 9]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+			intf, err := net.InterfaceByName(routes[i].Interface)
+			if err != nil {
+				glog.Infof("Can't find an interface named %q: %v\n", routes[i].Interface, err)
```
Comment (thockin) [inline pkg/util/util.go]: Warningf

File: pkg/util/util.go [hunk 10]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+			}
+			if intf.Flags&net.FlagUp == 0 {
+				glog.Infof("Interface %q is not up", routes[i].Interface)
```
Comment (thockin) [inline pkg/util/util.go]: V(3)

File: pkg/util/util.go [hunk 11]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+			}
+			addrs, err := intf.Addrs()
+			if err != nil {
+				glog.Infof("Can't get addresses for interface %q: %v\n", routes[i].Interface, err)
```
Comment (thockin) [inline pkg/util/util.go]: Warningf

File: pkg/util/util.go [hunk 12]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+				continue
+			}
+			if len(addrs) == 0 {
+				glog.Infof("Interface %q has no address", routes[i].Interface)
```
Comment (thockin) [inline pkg/util/util.go]: Warningf

File: pkg/util/util.go [hunk 13]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+				ip, _, err := net.ParseCIDR(addrs[i].String())
```
Comment (thockin) [inline pkg/util/util.go]: This is an out-of-bounds reference - you just checked that len(addrs) == 0

File: pkg/util/util.go [hunk 14]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+				if err != nil {
+					glog.Infof("Can't get addresses for interface %q: %v\n", addrs[i].String(), err)
+				}
+				finalIP = ip
+				found = true
+
+			} else if len(addrs) > 1 {
+				glog.Infof("Interface %q has %d addresses. %v\n", routes[i].Interface, len(addrs), addrs)
```
Comment (thockin) [inline pkg/util/util.go]: V(3)

File: pkg/util/util.go [hunk 15]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+				for i := range addrs {
+					glog.Infof("Checking interface %s . \n", addrs[i].String())
```
Comment (thockin) [inline pkg/util/util.go]: V(3)

File: pkg/util/util.go [hunk 16]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+					ip, _, err := net.ParseCIDR(addrs[i].String())
+					if err != nil {
+						continue
+					}
+					if ip.To4() == nil {
```
Comment (thockin) [inline pkg/util/util.go]: Add a TODO: to support v6

File: pkg/util/util.go [hunk 17]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+						glog.Infof("%v is not a valid IPv4 address\n", ip)
+						continue
+
+					}
+					if ip.IsLoopback() {
+						glog.Infof("Loopback found ", ip.String())
```
Comment (thockin) [inline pkg/util/util.go]: V(3)

File: pkg/util/util.go [hunk 18]
```diff
@@ -229,3 +233,128 @@ func SplitQualifiedName(str string) (string, string) {
...
+				}
+			}
+		}
+		if found {
+			break
+		}
+	}
+
+	if found {
+		glog.Infof("Choosing IP %v \n", finalIP)
```
Comment (thockin) [inline pkg/util/util.go]: V(3)

**Comment by thockin** [comment]
Thanks for integrating my sketch and fleshing it out.  Can I trouble you for some tests with varying "route -n" results that tickle the various modes?

Once this is in, we can switch the proxy to use it, too.

**Comment by sub-mod** [comment]
@brendanburns @thockin Kindly review

**Comment by thockin** [comment]
rebase?

**Comment by sub-mod** [comment]
@thockin rebased . Not sure why build fails . It works for me.

**Comment by thockin** [comment]
Change LGTM.

Shippable says:

Auto-merging pkg/master/master.go
CONFLICT (content): Merge conflict in pkg/master/master.go
Automatic merge failed; fix conflicts and then commit the result.

**Comment by sub-mod** [comment]
@thockin does it look ok now?

**Pull request merged**